################################################################################
SHORT DESCRIPTION:
################################################################################
Libxl support for passing a crypto key directory for a guest VM's disk(s).
If absent, defaults to an OpenXT-specific "/config/platform-crypto-keys".

################################################################################
LONG DESCRIPTION:
################################################################################
Plumb "crypto_key_dir" through the toolstack to set TAPDISK3_CRYPTO_KEYDIR env.

################################################################################
CHANGELOG
################################################################################
Authors:
Chris Rogers <rogersc@ainfosec.com>

Revised by:
Christopher Clark <christopher.clark6@baesystems.com>

################################################################################
REMOVAL
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################

* The hardcoded filesystem path "/config/platform-crypto-keys" will need to be
  removed from this libxl patch.

* Review whether a crypto key directory per disk, rather than just per VM, is
  the correct design.

  The device_disk struct is used widely within libxl for various types of disks,
  so modifying it requires care to ensure that all other sections of code
  using it continue to be correct.
    eg. This patch does not add support for reading the new crypto_key_dir data
        member from a serialized disk config, which may be inconsistent with
        the way other members of the struct are managed.

  Adding crypto_key_dir to the disk struct is, however, a convenient way of
  plumbing the data through to the site where it is used.

################################################################################
INTERNAL DEPENDENCIES
################################################################################

################################################################################
PATCHES
################################################################################
--- a/tools/libxl/libxl_internal.h
+++ b/tools/libxl/libxl_internal.h
@@ -1771,7 +1771,8 @@ _hidden int libxl__blktap_enabled(libxl_
  */
 _hidden char *libxl__blktap_devpath(libxl__gc *gc,
                                     const char *disk,
-                                    libxl_disk_format format);
+                                    libxl_disk_format format,
+                                    char *keydir);
 
 _hidden int libxl__get_tap_minor(libxl__gc *gc, libxl_disk_format format, const char *disk);
 _hidden bool tapdev_is_shared(libxl__gc *gc, const char *params);
--- a/tools/libxl/libxl_types.idl
+++ b/tools/libxl/libxl_types.idl
@@ -487,6 +487,9 @@ libxl_domain_build_info = Struct("domain
     ("blkdev_start",    string),
     ("cores_per_socket",  integer),
 
+    # directory containing the crypto keys for the VM's disks
+    ("crypto_key_dir", string),
+
     ("vnuma_nodes", Array(libxl_vnode_info, "num_vnuma_nodes")),
     
     ("device_model_version", libxl_device_model_version),
@@ -653,7 +656,8 @@ libxl_device_disk = Struct("device_disk"
     ("colo_port", integer),
     ("colo_export", string),
     ("active_disk", string),
-    ("hidden_disk", string)
+    ("hidden_disk", string),
+    ("crypto_key_dir", string)
     ])
 
 libxl_device_nic = Struct("device_nic", [
--- a/tools/xl/xl_parse.c
+++ b/tools/xl/xl_parse.c
@@ -1053,6 +1053,7 @@ void parse_config_data(const char *confi
     if (!xlu_cfg_get_long(config, "max_event_channels", &l, 0))
         b_info->event_channels = l;
 
+    xlu_cfg_replace_string (config, "crypto_key_dir", &b_info->crypto_key_dir, 0);
     xlu_cfg_replace_string (config, "kernel", &b_info->kernel, 0);
     xlu_cfg_replace_string (config, "ramdisk", &b_info->ramdisk, 0);
     xlu_cfg_replace_string (config, "device_tree", &b_info->device_tree, 0);
@@ -1395,6 +1396,13 @@ void parse_config_data(const char *confi
             parse_disk_config(&config, buf2, disk);
 
             free(buf2);
+
+            if (d_config->b_info.crypto_key_dir) {
+                replace_string(&disk->crypto_key_dir,
+                               d_config->b_info.crypto_key_dir);
+            } else {
+                disk->crypto_key_dir = NULL;
+            }
         }
     }
 
--- a/tools/libxl/libxl_blktap3.c
+++ b/tools/libxl/libxl_blktap3.c
@@ -64,7 +64,8 @@ int libxl__blktap_enabled(libxl__gc *gc)
 }
 
 char *libxl__blktap_devpath(libxl__gc *gc, const char *disk,
-		libxl_disk_format format)
+		libxl_disk_format format,
+		char *keydir)
 {
     const char *type = NULL;
     char *params, *devname = NULL;
@@ -86,7 +87,12 @@ char *libxl__blktap_devpath(libxl__gc *g
 
 	/* TODO Should we worry about return codes other than ENOENT? */
 
-    params = libxl__sprintf(gc, "%s:%s", type, disk);
+    if (!keydir || !strncmp(keydir, "", 1))
+        setenv("TAPDISK3_CRYPTO_KEYDIR", "/config/platform-crypto-keys", 1);
+    else
+        setenv("TAPDISK3_CRYPTO_KEYDIR", keydir, 1);
+
+    params = GCSPRINTF("%s:%s", type, disk);
 
     err = tap_ctl_create(params, &devname, flags, -1, 0, 0);
     if (!err) {
--- a/tools/libxl/libxl_dm.c
+++ b/tools/libxl/libxl_dm.c
@@ -1502,7 +1502,7 @@ static int libxl__build_device_model_arg
                  */
                 if (disks[i].backend == LIBXL_DISK_BACKEND_TAP)
                     target_path = libxl__blktap_devpath(gc, disks[i].pdev_path,
-                                                        disks[i].format);
+                                                        disks[i].format, disks[i].crypto_key_dir);
                 else
                     target_path = libxl__device_disk_find_local_path(gc,
                                                  guest_domid, &disks[i], true);
--- a/tools/libxl/libxl_disk.c
+++ b/tools/libxl/libxl_disk.c
@@ -322,7 +322,8 @@ static void device_disk_add(libxl__egc *
 
             case LIBXL_DISK_BACKEND_TAP:
                 dev = libxl__blktap_devpath(gc, disk->pdev_path,
-                                            disk->format);
+                                            disk->format,
+                                            disk->crypto_key_dir);
                 if (!dev) {
                      LOGD(ERROR, domid, "failed to get blktap devpath for %s: %s\n",
                           disk->pdev_path, strerror(rc));
